simplecpp.cpp

```diff
@@ -2302,15 +2302,192 @@ static void simplifyName(simplecpp::TokenList &expr)
     }
 }
 
+/*
+ * Reads at least minlen and at most maxlen digits (inc. prefix) in base base
+ * from s starting at position pos and converts them to a 
+ * unsigned long long value, updating pos to point to the first
+ * unused element of s.
+ * Returns ULLONG_MAX if the result is not representable and
+ * throws if the above requirements were not possible to satisfy.
+ */
+static unsigned long long stringToULLbounded(
+        const std::string& s,
+        std::size_t& pos,
+        int base = 0,
+        std::ptrdiff_t minlen = 1,
+        std::size_t maxlen = std::string::npos
+) {
+    std::string sub = s.substr(pos, maxlen);
+    const char* start = sub.c_str();
+    char* end;
+    unsigned long long value = std::strtoull(start, &end, base);
+    pos += end - start;
+    if(end - start < minlen)
+        throw std::runtime_error("expected digit");
+    return value;
+}
+
+/* Converts character literal (including prefix, but not ud-suffix)
+ * to long long value.
+ *
+ * Assumes ASCII(-only) encoded tok.
+ * 
+ * For target assumes
+ * - CHAR_BIT == 8
+ * - UTF-8 execution character set
+ * - UTF-32 execution wide-character set
+ * - requirements for __STDC_UTF_16__, __STDC_UTF_32__ and __STDC_ISO_10646__ satisfied
+ * - char16_t is 16bit wide
+ * - char32_t is 32bit wide
+ * - wchar_t is 32bit wide and unsigned
+ * - matching char signedness to host
+ * - matching width and representation of int to host
+ *
+ * For host assumes
+ * - two's complement
+ * - ASCII-compatible execution character set
+ *
+ * Implements multi-character narrow literals according to GCC's behavior,
+ * except multi code unit universal character names are not supported.
+ * Multi-character wide literals are not supported.
+ */
+static long long characterLiteralToLL(const std::string& tok)
+{
+    // default is wide/utf32
+    bool narrow = false;
+    bool utf8 = false;
+    bool utf16 = false;
+    
+    std::size_t pos;
+
+    if(tok[0] == '\'') {
+        narrow = true;
+        pos = 1;
+    } else if(tok[0] == 'u' && tok[1] == '\'') {
+        utf16 = true;
+        pos = 2;
+    } else if(tok[0] == 'u' && tok[1] == '8' && tok[2] == '\'') {
+        utf8 = true;
+        pos = 3;
+    } else if((tok[0] == 'L' || tok[0] == 'U') && tok[1] == '\'') {
+        pos = 2;
+    } else
+        throw std::runtime_error("expected a character literal");
+
+    unsigned long long multivalue = 0;
+
+    std::size_t nbytes = 0;
+
+    // multi-character literals are limited to the width of int
+    while(tok[pos] && tok[pos] != '\'') {
+        if(nbytes >= 1 && !narrow)
+            throw std::runtime_error("multiple characters only supported in narrow character literals");
+
+        unsigned long long value;
+        
+        if(tok[pos++] == '\\') {
```

Comment:
I can do this, but I will need to increment in both branches anyway. It would just be at the top in the first branch and at the bottom in the second branch.

Reply:
I still think that would be better.. thanks!